### PR TITLE
feat(bloc)!: v8.0.0

### LINKED ISSUE #2278: BlocUnhandledErrorException in assert does not get fired in release mode
Hi,

I was having some issues with my flutter app because of an exception not being logged on sentry when in release mode, so I did come digging in the `flutter_bloc` code and found that the `BlocUnhandledErrorException` is thrown only in debug mode [because it is wrapped in an assert](https://github.com/felangel/bloc/blob/97f826273182f738a14224ca670ebc96c3045e6d/packages/bloc/lib/src/bloc.dart#L388).

I'm not sure if it's a desired behavior or not, but I think that letting this throw in release mode too would improve unhandled exceptions logging on a release app, or a release-compiled dart code in general, I guess, since this is not `flutter_bloc`.

If this is a desired behavior, may I ask why? And most importantly, my workaround was to "catch" the error via a `BlocObserver` and reporting the error to sentry from there... is there a better solution?

Thank you!

Antonello

### PR DESCRIPTION
**Pull request by felangel**
## Status

**READY**

## Breaking Changes

YES

## Description

- feat(bloc)!: v8.0.0

#### package:bloc

- **BREAKING**: feat: BlocOverrides API
  - `Bloc.observer` removed in favor of `BlocOverrides.runZoned` and `BlocOverrides.current.blocObserver`
  - `Bloc.transformer` removed in favor of `BlocOverrides.runZoned` and `BlocOverrides.current.eventTransformer`
- **BREAKING**: refactor: `BlocObserver` is abstract and cannot be instantiated
- **BREAKING**: feat: `add` throws `StateError` when bloc is closed ([#2912](https://github.com/felangel/bloc/pull/2912))
- **BREAKING**: feat: `emit` throws `StateError` when bloc is closed ([#2913](https://github.com/felangel/bloc/pull/2913))
- **BREAKING**: feat: improve error handling/reporting
  - closes #2278
- **BREAKING**: feat: restrict scope of `emit` in `Bloc` and `Cubit`
  - In `Cubit`, `emit` is `protected` so it can only be used within the `Cubit` instance.
  - In `Bloc`, `emit` is `internal` so it cannot be used outside of the internal package implementation.
- **BREAKING**: refactor: remove deprecated `TransitionFunction`
- **BREAKING**: refactor: remove deprecated `transformEvents`
- **BREAKING**: refactor: remove deprecated `mapEventToState`
- **BREAKING**: refactor: remove deprecated `transformTransitions`
- **BREAKING**: refactor: remove deprecated `listen` on `BlocBase`
- feat: throw `StateError` if an event is added without a registered event handler

#### package:bloc_concurrency

- feat: upgrade to `bloc v8.0.0-dev.5`

#### package:bloc_test

- **BREAKING**: feat: upgrade to `bloc v8.0.0-dev.5`
- **BREAKING**: feat: upgrade to `mocktail v0.2.0`
- feat: `MockBloc` no longer implicitly requires `registerFallbackValue` for events and states

#### package:hydrated_bloc

- **BREAKING**: feat: introduce `HydratedBlocOverrides` API ([#2947](https://github.com/felangel/bloc/pull/2947))
  - `HydratedBloc.storage` removed in favor of `HydratedBlocOverrides.runZoned` and `HydratedBlocOverrides.current.storage`
- **BREAKING**: feat: upgrade to `bloc v8.0.0-dev.5`

#### package:angular_bloc

- **BREAKING**: feat: upgrade to `bloc v8.0.0-dev.5`

#### package:flutter_bloc

- **BREAKING**: feat: upgrade to `bloc v8.0.0-dev.5`

#### package:replay_bloc

- **BREAKING**: feat: upgrade to `bloc v8.0.0-dev.5`

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [X] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

**Comment by felangel** [comment]
Published [bloc v8.0.0-dev.1](https://pub.dev/packages/bloc/versions/8.0.0-dev.1) 🎉

**Comment by felangel** [comment]
Published [bloc v8.0.0-dev.2](https://pub.dev/packages/bloc/versions/8.0.0-dev.2) 🎉

**Comment by felangel** [comment]
Published [bloc_test v9.0.0-dev.1](https://pub.dev/packages/bloc_test/versions/9.0.0-dev.1) 🎉

**Comment by PiN73** [comment]
Are you really going to break core API?

**Comment by felangel** [comment]
> Are you really going to break core API?

Yup bloc v7.2.0 deprecates the various APIs that will be removed in v8.0.0 so you have time to migrate gradually. See https://github.com/felangel/bloc/issues/2526 for more info.

**Comment by felangel** [comment]
Published [hydrated_bloc v8.0.0-dev.1](https://pub.dev/packages/hydrated_bloc/versions/8.0.0-dev.1) 🎉

**Comment by felangel** [comment]
Published [flutter_bloc v8.0.0-dev.1](https://pub.dev/packages/flutter_bloc/versions/8.0.0-dev.1) 🎉

**Comment by felangel** [comment]
Published [sealed_flutter_bloc v8.0.0-dev.1](https://pub.dev/packages/sealed_flutter_bloc/versions/8.0.0-dev.1) 🎉

**Comment by felangel** [comment]
Published [bloc v8.0.0-dev.3](https://pub.dev/packages/bloc/versions/8.0.0-dev.3) 🎉

**Comment by felangel** [comment]
Published [bloc v8.0.0-dev.4](https://pub.dev/packages/bloc/versions/8.0.0-dev.4) 🎉

**Comment by felangel** [comment]
Published [bloc_test v9.0.0-dev.4](https://pub.dev/packages/bloc_test/versions/9.0.0-dev.4) 🎉

**Comment by felangel** [comment]
Published [hydrated_bloc v8.0.0-dev.3](https://pub.dev/packages/hydrated_bloc/versions/8.0.0-dev.3) 🎉

**Comment by cedvdb** [comment]
how close are we to a stable release ?

**Comment by felangel** [comment]
> how close are we to a stable release ?

Pretty close. I’m hoping to have a stable release within the next week or two.

File: packages/bloc/lib/src/bloc.dart [hunk 1]
```diff
@@ -207,46 +206,10 @@ Please make sure to await all asynchronous operations within event handlers.
...
-  /// Stack trace which accompanied the error.
-  /// May be [StackTrace.empty] if no stack trace was provided.
-  final StackTrace stackTrace;
-
-  @override
-  String toString() {
-    return 'Unhandled error $error occurred in $bloc.\n'
-        '$stackTrace';
-  }
+class _Handler {
```
Comment (lhmzhou) [inline packages/bloc/lib/src/bloc.dart:209]: Curious: why is this handler made private?

File: packages/bloc/test/bloc_on_test.dart [hunk 2]
```diff
@@ -62,6 +62,16 @@ void main() {
...
     });
 
+    test('throws StateError when handler is missing', () {
+      const expectedMessage =
+          '''add(TestEventA) was called without a registered event handler.\n'''
+          '''Make sure to register a handler via on<TestEventA>((event, emit) {...})''';
+      final expected = throwsA(
+        isA<StateError>().having((e) => e.message, 'message', expectedMessage),
+      );
+      expect(() => MissingHandlerBloc()..add(TestEventA()), expected);
```
Comment (lhmzhou) [inline packages/bloc/test/bloc_on_test.dart:72]: Can you explain the cascade notation here? What does the second dot notation return?

File: packages/bloc/lib/src/bloc.dart (hunk 1 above)
Comment (felangel) [reply packages/bloc/lib/src/bloc.dart:209]: It's only used internally so there's no need to expose it publicly.

File: packages/bloc/test/bloc_on_test.dart (hunk 2 above)
Comment (felangel) [reply packages/bloc/test/bloc_on_test.dart:72]: Check out https://dart.dev/guides/language/language-tour#cascade-notation.

Comment (felangel) [reply packages/bloc/test/bloc_on_test.dart]: Good catch, the cascade operator is unnecessary in this case 👍

**Review by lhmzhou — APPROVED** [review]
:shipit:

**Comment by felangel** [comment]
Published [bloc v8.0.0-dev.5](https://pub.dev/packages/bloc/versions/8.0.0-dev.5) 🎉

**Comment by felangel** [comment]
Published [angular_bloc v8.0.0-dev.3](https://pub.dev/packages/angular_bloc/versions/8.0.0-dev.3) 🎉

**Comment by felangel** [comment]
Published [bloc_test v9.0.0-dev.5](https://pub.dev/packages/bloc_test/versions/9.0.0-dev.5) 🎉

**Comment by felangel** [comment]
Published [bloc_concurrency v0.2.0-dev.2](https://pub.dev/packages/bloc_concurrency/versions/0.2.0-dev.2) 🎉

**Comment by felangel** [comment]
Published [replay_bloc v0.2.0-dev.3](https://pub.dev/packages/replay_bloc/versions/0.2.0-dev.3) 🎉

**Comment by felangel** [comment]
Published [flutter_bloc v8.0.0-dev.3](https://pub.dev/packages/flutter_bloc/versions/8.0.0-dev.3) 🎉

**Comment by felangel** [comment]
Published [hydrated_bloc v8.0.0-dev.4](https://pub.dev/packages/hydrated_bloc/versions/8.0.0-dev.4) 🎉

**Comment by felangel** [comment]
Published [hydrated_bloc v8.0.0-dev.5](https://pub.dev/packages/hydrated_bloc/versions/8.0.0-dev.5) 🎉

**Pull request merged**